### PR TITLE
WIP Model card

### PR DESCRIPTION
**Pull request by merveenoyan**
This is a WIP for model card. Opening this PR so you could take a quick look to see if the behavior seems good. 
I will add what I can add on top of it (Evaluation results, model plot and so on)

### CODE REVIEW

**Comment by merveenoyan** [comment]
I checked out from my documentation PR, feel free to ignore the parts about adding docstrings.

File: training_pipeline/pipeline/components/HFPusher/model_card.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def create_card(card_data, template_path, model_metadata, 
+                placeholders, **template_kwargs):
+    """Creates model card.
+
+    Args:
+        model_metadata (dict): Dict of card metadata.
+        see here for what you can pass to the metadata section:
+        https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/templates/modelcard_template.md
```
Comment (sayakpaul) [inline training_pipeline/pipeline/components/HFPusher/model_card.py:11]:
```suggestion
        Refer to the link to know what you can pass to the metadata section:
        https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/templates/modelcard_template.md
```

More inclusive. Not everyone can see.

File: training_pipeline/pipeline/components/HFPusher/model_card.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+from huggingface_hub import ModelCard, ModelCardData
+
+
+def create_card(card_data, template_path, model_metadata, 
+                placeholders, **template_kwargs):
+    """Creates model card.
+
+    Args:
+        model_metadata (dict): Dict of card metadata.
+        see here for what you can pass to the metadata section:
```
Comment (sayakpaul) [inline training_pipeline/pipeline/components/HFPusher/model_card.py:10]: Include descriptions about the other arguments too.

File: training_pipeline/pipeline/components/HFPusher/model_card.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/templates/modelcard_template.md
+
+    Returns:
+        Model card (ModelCard): instantiated and filled ModelCard.
+    """
+    if model_metadata is None:
+        model_metadata = {}
+    model_metadata["library_name"] = "vertex"
```
Comment (sayakpaul) [inline training_pipeline/pipeline/components/HFPusher/model_card.py:18]: If one runs the pipeline on a different orchestrator then `vertex` won't make sense. Also, it's better to call the key `orchestrator` or `runner` to denote that it's the platform on which things are getting executed.

File: training_pipeline/pipeline/components/HFPusher/model_card_template.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+---
+# For reference on model card metadata, see: https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1
```
Comment (sayakpaul) [inline training_pipeline/pipeline/components/HFPusher/model_card_template.md:2]: see => refer.

File: training_pipeline/pipeline/components/HFPusher/model_card_template.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+{{card_data}}
+---
+
+# Model Card for {{ model_id | default("Model ID", true) }}
+
+# Model Details
+
+## Model Description
+
+This is a {{ task | default("segmentation")}} model trained using Vertex AI.
```
Comment (sayakpaul) [inline training_pipeline/pipeline/components/HFPusher/model_card_template.md:12]: If we're not using Vertex AI, then it doesn't make sense.

File: training_pipeline/pipeline/components/HFPusher/model_card_template.md (hunk 4 above)
Comment (sayakpaul) [inline training_pipeline/pipeline/components/HFPusher/model_card_template.md:2]: Does this link also show how we can specify the formatting like:

```
This is a {{ task | default("segmentation")}} model trained using Vertex AI.
```

**Review by sayakpaul — CHANGES_REQUESTED** [review]
Looking good so far. 

Did you use the model card utility as a part of the pipeline yet?

No worries, if not. I was just being curious.

File: training_pipeline/pipeline/components/HFPusher/model_card_template.md (hunk 4 above)
Comment (merveenoyan) [reply training_pipeline/pipeline/components/HFPusher/model_card_template.md:2]: it just shows the keys, so nope.

File: training_pipeline/pipeline/components/HFPusher/model_card_template.md (hunk 5 above)
Comment (merveenoyan) [reply training_pipeline/pipeline/components/HFPusher/model_card_template.md:12]: What is this project specifically made for?

File: training_pipeline/pipeline/components/HFPusher/model_card_template.md (hunk 4 above)
Comment (merveenoyan) [reply training_pipeline/pipeline/components/HFPusher/model_card_template.md:2]: Thanks!

File: training_pipeline/pipeline/components/HFPusher/model_card_template.md (hunk 5 above)
Comment (sayakpaul) [reply training_pipeline/pipeline/components/HFPusher/model_card_template.md:12]: The project can be run any compatible platform. We ran it locally and Vertex AI. 

Cc: @deep-diver

Comment (deep-diver) [reply training_pipeline/pipeline/components/HFPusher/model_card_template.md:12]: yes it is. the pipeline has two targets, Local and Vertex AI (which is Kubeflow 2.x)

**Comment by deep-diver** [comment]
It looks good as a starting point. I just want to ask to add a notebook that shows before and after running `create_card()` function to clearly see what the generated model card looks like. WDYT?

**Comment by merveenoyan** [comment]
@deep-diver sure I'll do that 🙂

**Comment by deep-diver** [comment]
@merveenoyan thanks! that would be great 👍🏼

**Comment by sayakpaul** [comment]
@deep-diver 

@merveenoyan is finishing the first version of the model card feature in this PR. She wanted to test `HFPusher` (which has the model card utility) in isolation, preferably in a notebook. 

Do you have any starter notebook which she could use?

**Comment by merveenoyan** [comment]
The revision of the test repository lies here, I tested like below: https://huggingface.co/merve/tfx-blessed-model/blob/merve/README.md

```python
import tensorflow as tf
from runner import deploy_model_for_hf_hub

model = model_init() # random dummy model
model.save("./model_save_dir")

model_card_metadata = {}
deploy_model_for_hf_hub(
    username="merve",
    access_token=access_token,
    repo_name="tfx-blessed-model",
    model_path="./model_save_dir",
    model_version="merve"
)
```

**Comment by merveenoyan** [comment]
@deep-diver had to add `create_card` to `__init__ `so it was importable :/

**Review by sayakpaul — APPROVED** [review]
Thanks a lot for working on this! 

Will defer to @deep-diver for the final merge.

**Comment by merveenoyan** [comment]
Closes #43

**Pull request merged**